### DISCUSSION TITLE
Long Term Architecture Vision

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#1635] [Ideas]
# Long Term Architecture Vision

> **Warning**
> Early draft, work in progress

This document is intended to capture long term (5+ years) architectural designs.  The goal is to create alignment about "where are we going" and "what is our North Star?"  This is a high level document, which will link out to more detailed documents about the various subsystems.

## End-to-end features and capabilities

- [FHIR Terminology service](https://github.com/medplum/medplum/discussions/1411)
- FHIR Profiles
- CDS Hooks
- DICOM
- Provisioned VPCs and VPNs
- Provisioned SFTP / S3
- Provisioned "apps" sites using CloudFront / S3 / Vercel
- Bot marketplace
  - First and third party bots for specialty integrations and capabilities
- Enterprise Features
  - Security management
  - Compliance management

## Auth and user management

- [Extended access policies](https://github.com/medplum/medplum/discussions/1453)
- [Project roles/groups](https://github.com/medplum/medplum/discussions/1392)

## Back end architecture

- [Plugin architecture](https://github.com/medplum/medplum/discussions/1424)
- [Websockets](https://github.com/medplum/medplum/discussions/1490)
- [Referential integrity](https://github.com/medplum/medplum/discussions/1221)
- [Client assigned IDs](https://github.com/medplum/medplum/discussions/1175)

### High availability and scalability

- Multi-Region Infrastructure Deployment
- Beyond Postgres (DynamoDB, Cassandra, ScyllaDB, Typesense, ElasticSearch)
  - ScyllaDB scales to trillions of rows: https://discord.com/blog/how-discord-stores-trillions-of-messages
  - ElasticSearch scales to billions of documents per index: https://stackoverflow.com/questions/61471665/max-number-of-records-in-elasticsearch

## Front end architecture

The main challenge with all front end planning is to find the balance between:
1. Usability
2. Aesthetics
3. Performance
4. Developer experience
5. Futureproofing

- Official design system
- Beyond React (WebComponents)

## Dev ops and infra

- [Publishing Debian packages](https://github.com/medplum/medplum/discussions/1577)

## Testing and quality assurance

- End-to-end testing via Playwright or Cypress
- More automation in short lived staging environments for deployments via AWS CodePipeline

## Accessibility (a11y), Internationalization (i18n), and Localization (l10n)

- Per-locale metadata resources (StructureDefinition, SearchParameter, CodeSystem, ValueSet, etc)
- Continuous localization

**Comment by joekendal** [comment]
I'm liking the sound of DynamoDB & OpenSearch support.

**Comment by joekendal** [comment]
The Bot marketplace sounds great
